### DISCUSSION TITLE
@rsuite-icons and Content-Security-Policy header

### DISCUSSION THREAD

**Discussion by obdulia-losantos** [rsuite/rsuite#3997] [General]
I have a project that requires to add CSP header but the icons lib is injecting the styles directly to the document.

https://github.com/rsuite/rsuite-icons/blob/master/src/Icon.tsx#L82

```typescript
import { useClassNames, useInsertStyles } from './utils';

const Icon = React.forwardRef<SVGElement, IconProps>(
  // ...
  
    useInsertStyles();
```

- `utils/useInsertStyles.ts`

```typescript
import { insertCss } from 'insert-css';

// ...

const useInsertStyles = (styleStr = styles) => {
  useEffect(() => {
    // Make sure css injected once.
    if (!cssInjected) {
      insertCss(styleStr, { prepend: true });
      cssInjected = true;
    }
  }, []);
};
```

This is the error I have

```text
Refused to apply inline style because it violates the following Content Security Policy directive: "style-src 'self'. Either the 'unsafe-inline' keyword, a hash ('sha256-zMJRSletExaG2OrRStnb8KbALyvN1u/BZKxAMeIr0gQ='), or a nonce ('nonce-...') is required to enable inline execution.
```

Is it possible to drop that behavior and import the styles like any other dependency like `rsuite-table` suggest?

```typescript
import { Table, Column, HeaderCell, Cell } from 'rsuite-table';
import 'rsuite-table/lib/less/index.less'; // or 'rsuite-table/dist/css/rsuite-table.css'
```

**Comment by simonguo** [comment]
Thank you for your feedback, we will consider a solution to solve the problem.

https://github.com/rsuite/rsuite-icons/pull/12

**Reply by simonguo** [reply]
In the [latest version](https://github.com/rsuite/rsuite/releases/tag/v5.73.0), an `csp` property is added to CustomProvider, which can configure the [nonce][nonce] value of [Content Security Policy][csp].

```jsx
import { CustomProvider } from 'rsuite';

return (
  <CustomProvider csp={{ nonce: 'xxxxxx' }}>
    <App />
  </CustomProvider>
);
```

If you want to disable inline styles, you can add a `disableInlineStyles` property to CustomProvider.

```jsx
import '@rsuite/icons/dist/css/rsuite-icon.css';

return (
  <CustomProvider disableInlineStyles>
    <App />
  </CustomProvider>
);
```

[csp]: https://developer.mozilla.org/zh-CN/docs/Web/HTTP/CSP
[nonce]: https://developer.mozilla.org/zh-CN/docs/Web/HTML/Global_attributes/nonce

**Comment by obdulia-losantos** [comment]
Now, I have the same problem with RSuite v6. I cannot upgrade because the csp value is not propagated to the `CustomProvider`, only to the `IconProvider`.

This is the fix: https://github.com/rsuite/rsuite/pull/4499
